tiledb/sm/query/readers/reader_base.cc

```diff
@@ -1010,6 +1011,66 @@ Status ReaderBase::zip_tile_coordinates(
   return Status::Ok();
 }
 
+template<typename ResultTileType, typename ResultCoordsType>
+void ReaderBase::compute_hilbert_values(const std::vector<ResultTile*>& result_tiles) const {
+  auto timer_se = stats_->start_timer("compute_hilbert_values");
+
+  // For easy reference.
+  auto dim_num = array_schema_.dim_num();
+
+  // Create a Hilbert class.
+  Hilbert h(dim_num);
+  auto bits = h.bits();
+  auto max_bucket_val = ((uint64_t)1 << bits) - 1;
+
+  // Parallelize on tiles.
+  auto status = parallel_for(
+      storage_manager_->compute_tp(), 0, result_tiles.size(), [&](uint64_t t) {
+        /*
+        if constexpr (std::is_same<ResultCoords, ResultCoordsType>::value) {
+          result_tiles[t]->set_zipped();
+        } */
+       auto tile =
+            static_cast<ResultTileType*>(result_tiles[t]);
+  
+      size_t cell_num = result_tiles[t]->cell_num();
+      if constexpr (!std::is_same<ResultCoords, ResultCoordsType>::value) {
+        cell_num = fragment_metadata_[tile->frag_idx()]->cell_num(tile->tile_idx());
+      }
+      auto rc = ResultCoordsType(tile, 0);
+        std::vector<uint64_t> coords(dim_num);
+        result_tiles[t]->allocate_hilbert_vector();
+        for (rc.pos_ = 0; rc.pos_ < cell_num; rc.pos_++) {
+          // Process only values in bitmap.
```

Comment:
Note that in some cases, adding a conditional in a for loop could break vectorization. But in this case, the Hilbert computation is already not vectorized so we are ok.

Reply:
We would also have to add a bitmap() function in order for this to work--currently I have a hacked (and honestly, not great imo) solution in place for this where I use the following conditional so that I don't have to add a bitmap() function to the ResultTile class. I prefer to not add one since it seems unreasonable that a bitmap() function would exist for a class with no valid bitmap. Perhaps we could do something with optionals? But that might require a more comprehensive refactoring.
```
bool compute_cond = !tile->has_bmp();
if constexpr (!std::is_same<ResultCoords, ResultCoordsType>::value) {
  compute_cond = compute_cond || tile->bitmap()[rc.pos_];
}
```
This at least compiles to my knowledge. Clean code should be a priority but I don't know whether this is the optimal solution.